File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/processors/gbd/DbContextLoader.java
package ru.kmorozov.library.data.loader.processors.gbd;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Component;
import ru.kmorozov.db.core.config.IContextLoader;
import ru.kmorozov.db.core.logic.model.book.BookInfo;
import ru.kmorozov.gbd.core.logic.model.book.base.AbstractPage;
import ru.kmorozov.gbd.core.logic.model.book.base.IPage;
import ru.kmorozov.library.data.repository.GoogleBooksRepository;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

@Component
public class DbContextLoader implements IContextLoader {

    private final Map<String, BookInfo> booksMap = new HashMap<>();

    @Autowired
    @Lazy
    private GoogleBooksRepository googleBooksRepository;

    @Override
    public void updateIndex() {

    }

    @Override
    public void updateContext() {

    }

    @Override
    public void updateBookInfo(final BookInfo bookInfo) {
        final BookInfo existBookInfo = this.googleBooksRepository.findByBookId(bookInfo.getBookId());
        if (existBookInfo != null)
            this.googleBooksRepository.delete(existBookInfo);

        this.googleBooksRepository.save(bookInfo);
    }

    @Override
    public BookInfo getBookInfo(final String bookId) {
        BookInfo info = this.booksMap.get(bookId);
        if (info == null) {
            info = this.googleBooksRepository.findByBookId(bookId);

            for (final IPage page : info.getPages().getPages())
                ((AbstractPage) page).setFileExists(true);

            this.booksMap.put(bookId, info);
        }
        return info;
    }

    @Override
    public Set<String> getBookIdsList() {
        return this.booksMap.keySet();
    }

    @Override
    public int getContextSize() {
        return this.booksMap.size();
    }

    @Override
    public void refreshContext() {

    }

    @Override
    public boolean isValid() {
        return false;
    }

}

File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/processors/gbd/DbContextLoader.kt
package ru.kmorozov.library.data.loader.processors.gbd

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Component
import ru.kmorozov.db.core.config.IContextLoader
import ru.kmorozov.db.core.logic.model.book.BookInfo
import ru.kmorozov.gbd.core.logic.model.book.base.AbstractPage
import ru.kmorozov.gbd.core.logic.model.book.base.IBookInfo
import ru.kmorozov.library.data.repository.GoogleBooksRepository
import java.util.*

@Component
class DbContextLoader : IContextLoader {

    private val booksMap = HashMap<String, BookInfo>()

    @Autowired
    @Lazy
    private lateinit var googleBooksRepository: GoogleBooksRepository

    override val bookIdsList: Set<String>
        get() = booksMap.keys

    override val contextSize: Int
        get() = booksMap.size

    override val isValid: Boolean
        get() = false

    override fun updateIndex() {

    }

    override fun updateContext() {

    }

    override fun updateBookInfo(bookInfo: BookInfo) {
        val existBookInfo = googleBooksRepository.findByBookId(bookInfo.bookId)
        if (existBookInfo != null)
            googleBooksRepository.delete(existBookInfo)

        googleBooksRepository.save(bookInfo)
    }

    override fun getBookInfo(bookId: String): IBookInfo {
        return booksMap.computeIfAbsent(bookId, {
            val info = googleBooksRepository.findByBookId(bookId)!!

            for (page in info.pages.pages)
                (page as AbstractPage).isFileExists = true

            info
        })
    }

    override fun refreshContext() {

    }

}
